jerry-core/ecma/operations/ecma-regexp-object.c

```diff
@@ -238,86 +231,233 @@ ecma_regexp_unicode_advance (const lit_utf8_byte_t **str_p, /**< reference to st
+
+  JERRY_ASSERT (ecma_is_value_true (status));
+
+  return new_object_p;
+} /* ecma_op_regexp_alloc */
 
 /**
- * RegExp object creation operation.
+ * Helper method for initializing an aready existing RegExp object.
  *
- * See also: ECMA-262 v5, 15.10.4.1
+ * @return void
+ */
+static void
+ecma_op_regexp_initialize (ecma_object_t *regexp_obj_p, /**< RegExp object */
+                           const re_compiled_code_t *bc_p, /**< bytecode */
+                           ecma_string_t *pattern_str_p, /**< pattern */
+                           uint16_t flags) /**< flags */
+{
+  ecma_extended_object_t *ext_obj_p = (ecma_extended_object_t *) regexp_obj_p;
+
+  if (JERRY_UNLIKELY (ext_obj_p->u.class_prop.u.value != JMEM_CP_NULL))
+  {
+    re_compiled_code_t *old_bc_p = ECMA_GET_INTERNAL_VALUE_POINTER (re_compiled_code_t,
+                                                                    ext_obj_p->u.class_prop.u.value);
+
+    ecma_bytecode_deref ((ecma_compiled_code_t *) old_bc_p);
+
+#if !ENABLED (JERRY_ES2015)
+    ecma_regexp_update_props (regexp_obj_p, pattern_str_p, flags);
+#endif /* !ENABLED (JERRY_ES2015) */
+  }
+#if !ENABLED (JERRY_ES2015)
+  else
+  {
+    ecma_regexp_create_props (regexp_obj_p, pattern_str_p, flags);
+  }
+#endif /* !ENABLED (JERRY_ES2015) */
+
+#if ENABLED (JERRY_ES2015)
+  JERRY_UNUSED (pattern_str_p);
+  JERRY_UNUSED (flags);
+#endif /* ENABLED (JERRY_ES2015) */
+
+  ECMA_SET_INTERNAL_VALUE_POINTER (ext_obj_p->u.class_prop.u.value, bc_p);
+} /* ecma_op_regexp_initialize */
+
+/**
+ * Method for creating a RegExp object from pattern.
  *
- * @return constructed RegExp object - if pattern and flags were parsed successfully
- *         error ecma value          - otherwise
+ * Note:
+ *      Allocation have to happen before invoking this function using ecma_op_regexp_alloc.
  *
- *         Returned value must be freed with ecma_free_value
+ * @return ecma_value_t
+ */
+ecma_value_t
+ecma_op_create_regexp_from_pattern (ecma_object_t *regexp_obj_p, /**< RegExp object */
+                                    ecma_value_t pattern_value, /**< pattern */
+                                    ecma_value_t flags_value) /**< flags */
+{
+  ecma_string_t *pattern_str_p = ecma_regexp_read_pattern_str_helper (pattern_value);
+  uint16_t flags = 0;
+
+  if (JERRY_UNLIKELY (pattern_str_p == NULL))
+  {
+    return ECMA_VALUE_ERROR;
+  }
+
+  if (!ecma_is_value_undefined (flags_value))
+  {
+    ecma_string_t *flags_str_p = ecma_op_to_string (flags_value);
+
+    if (JERRY_UNLIKELY (flags_str_p == NULL))
+    {
+      ecma_deref_ecma_string (pattern_str_p);
+      return ECMA_VALUE_ERROR;
+    }
+
+    ecma_value_t parse_flags_value = ecma_regexp_parse_flags (flags_str_p, &flags);
+    ecma_deref_ecma_string (flags_str_p);
+
+    if (ECMA_IS_VALUE_ERROR (parse_flags_value))
+    {
+      ecma_deref_ecma_string (pattern_str_p);
+      return parse_flags_value;
+    }
+
+    ecma_free_value (parse_flags_value);
```

Comment:
If I understand correctly then `ecma_regexp_parse_flags` can only return `VALUE_EMPTY` or an error. If so then this `free_value` call could be turned into an assert to check if the parse result is `VALUE_EMTPY`. (Just like on line `396`.)